File: src/pages/LoginPage.ts
import { expect, Page } from "@playwright/test";
import BasePage from "./BasePage";

export default class LoginPage extends BasePage {
  constructor(page: Page) {
    super(page);
  }

  //All the UI elements of the Page.
  userNameTextBox = async () => this.Page.$("[data-test='username']");
  passwordTextBox = async () => this.Page.$("[data-test='password']");
  submitButton = async () => this.Page.$("[data-test='login-button']");
  error_message = async () => this.Page.$("[data-test='error']");

  /**
    * static loginFlow
    @param userName:string, 
    @param password: string 
    */
    public async loginFlow(userName: string, password: string) {
    await (await this.userNameTextBox()).fill(userName);
    await (await this.passwordTextBox()).fill(password);
    await (await this.submitButton()).click();
  }

  public async clearLoginTextBoxes() {
    await (await this.userNameTextBox()).fill("");
    await (await this.passwordTextBox()).fill("");
  }

  /**
   * verify login success flow.
   */
   public async verifyLoginSuccess() {
    const isVisibleUserMenu = await this.Page.isVisible(".bm-burger-button");
    expect(isVisibleUserMenu).toBeTruthy();
  }

  public async verifyErrorMessage(message) {
    const isVisibleError = await this.Page.isVisible("[data-test='error']");
    expect(isVisibleError).toBeTruthy();
    const errorcontent = await this.Page.innerText("[data-test='error']");
    expect(errorcontent).toContain(message);
  }

  public async logoutFlow() {
    const menu_button =".bm-burger-button";
    await this.Page.click(menu_button);
    await this.Page.click('text="Logout"');
    expect(await this.Page.isVisible("text=Login")).toBeTruthy();
  }
}

File: python/pages/login_page.py
"""Page object for the Sauce Demo login screen."""

from __future__ import annotations

from playwright.sync_api import Page, expect

from .base_page import BasePage


class LoginPage(BasePage):
    def __init__(self, page: Page) -> None:
        super().__init__(page)
        # The app exposes stable `data-test` attributes; Playwright's default
        # test-id attribute is configured to `data-test` in conftest.py.
        self.username = page.get_by_test_id("username")
        self.password = page.get_by_test_id("password")
        self.login_button = page.get_by_test_id("login-button")
        self.error_message = page.get_by_test_id("error")

    def open(self) -> None:
        self.goto("/")
        expect(self.login_button).to_be_visible()

    def login(self, username: str, password: str) -> None:
        self.username.fill(username)
        self.password.fill(password)
        self.login_button.click()

    def expect_login_error(self, expected: str) -> None:
        expect(self.error_message).to_be_visible()
        expect(self.error_message).to_contain_text(expected)
